File: packages/recorder/ios/HybridClarionRecorder.swift
import Foundation
import NitroModules

public typealias StateClosure = (String) -> Void
public typealias AudioLevelClosure = (Double, Double) -> Void
public typealias ChunkClosure = (String, Double, Double, Double) -> Void
public typealias ErrorClosure = (NativeRecorderError) -> Void

public final class HybridClarionRecorder: HybridClarionRecorderSpec {
  private var session: RecorderSession?
  private var currentStateValue: String = "idle"

  private var nextListenerId: Int = 1
  private var stateListeners: [Int: StateClosure] = [:]
  private var audioLevelListeners: [Int: AudioLevelClosure] = [:]
  private var chunkListeners: [Int: ChunkClosure] = [:]
  private var errorListeners: [Int: ErrorClosure] = [:]

  private let listenerLock = NSLock()

  public var state: String { currentStateValue }

  public func prepare(config: NativeRecorderConfig) throws -> Promise<Void> {
    return Promise.async { [weak self] in
      guard let self else { return }
      let parsed = config.toSwiftConfig()
      let newSession = RecorderSession(config: parsed, callbacks: self)
      self.emit(state: "preparing")
      try await newSession.prepare()
      self.session = newSession
      self.emit(state: "ready")
    }
  }

  public func start() throws -> Promise<Void> {
    return Promise.async { [weak self] in
      guard let self else { return }
      guard let session = self.session else {
        throw RecorderError(code: "ENGINE_NOT_READY", message: "Call prepare() first")
      }
      self.emit(state: "starting")
      try session.start()
    }
  }

  public func pause() throws -> Promise<Void> {
    return Promise.async { [weak self] in
      try self?.session?.pause()
    }
  }

  public func resume() throws -> Promise<Void> {
    return Promise.async { [weak self] in
      try self?.session?.resume()
    }
  }

  public func stop() throws -> Promise<NativeRecorderResult> {
    return Promise.async { [weak self] in
      guard let self, let session = self.session else {
        throw RecorderError(code: "ENGINE_NOT_READY", message: "No session")
      }
      let output = try await session.stop()
      return output.toNative()
    }
  }

  public func discard() throws -> Promise<Void> {
    return Promise.async { [weak self] in
      await self?.session?.discard()
    }
  }

  public func release() throws -> Promise<Void> {
    return Promise.async { [weak self] in
      guard let self else { return }
      await self.session?.release()
      self.session = nil
      self.removeAllListeners()
    }
  }

  public func addStateListener(listener: @escaping StateClosure) throws -> Double {
    register { id in stateListeners[id] = listener }
  }

  public func addAudioLevelListener(listener: @escaping AudioLevelClosure) throws -> Double {
    register { id in audioLevelListeners[id] = listener }
  }

  public func addChunkListener(listener: @escaping ChunkClosure) throws -> Double {
    register { id in chunkListeners[id] = listener }
  }

  public func addErrorListener(listener: @escaping ErrorClosure) throws -> Double {
    register { id in errorListeners[id] = listener }
  }

  public func removeListener(id: Double) throws {
    listenerLock.lock(); defer { listenerLock.unlock() }
    let key = Int(id)
    stateListeners.removeValue(forKey: key)
    audioLevelListeners.removeValue(forKey: key)
    chunkListeners.removeValue(forKey: key)
    errorListeners.removeValue(forKey: key)
  }

  public func removeAllListeners() {
    listenerLock.lock(); defer { listenerLock.unlock() }
    stateListeners.removeAll()
    audioLevelListeners.removeAll()
    chunkListeners.removeAll()
    errorListeners.removeAll()
  }

  private func register(_ assign: (Int) -> Void) -> Double {
    listenerLock.lock(); defer { listenerLock.unlock() }
    let id = nextListenerId
    nextListenerId += 1
    assign(id)
    return Double(id)
  }

  private func emit(state: String) {
    currentStateValue = state
    snapshotListeners(stateListeners).forEach { $0(state) }
  }

  private func snapshotListeners<T>(_ map: [Int: T]) -> [T] {
    listenerLock.lock(); defer { listenerLock.unlock() }
    return Array(map.values)
  }
}

extension HybridClarionRecorder: RecorderCallbacks {
  func onState(_ state: String) {
    emit(state: state)
  }

  func onAudioLevel(rms: Double, peak: Double) {
    snapshotListeners(audioLevelListeners).forEach { $0(rms, peak) }
  }

  func onChunk(uri: String, startMs: Double, endMs: Double, sizeBytes: Double) {
    snapshotListeners(chunkListeners).forEach { $0(uri, startMs, endMs, sizeBytes) }
  }

  func onError(_ error: RecorderError) {
    currentStateValue = "error"
    let payload = NativeRecorderError(
      code: error.code,
      message: error.message,
      recoverable: error.recoverable
    )
    snapshotListeners(errorListeners).forEach { $0(payload) }
  }
}

private extension NativeRecorderConfig {
  func toSwiftConfig() -> RecorderConfig {
    return RecorderConfig(
      sampleRate: sampleRate,
      channels: UInt32(channels),
      bitDepth: UInt32(bitDepth),
      outputDirectory: outputDirectory,
      filenamePrefix: filenamePrefix,
      rotateAfterMs: rotateAfterMs,
      emitAudioLevel: emitAudioLevel,
      audioLevelIntervalMs: audioLevelIntervalMs,
      aacBitrate: Int(aacBitrate)
    )
  }
}

private extension RecorderOutput {
  func toNative() -> NativeRecorderResult {
    return NativeRecorderResult(
      uri: uri,
      durationMs: durationMs,
      sizeBytes: sizeBytes,
      sampleRate: sampleRate,
      channels: Double(channels),
      bitDepth: Double(bitDepth)
    )
  }
}

File: packages/recorder/android/src/main/java/com/margelo/nitro/clarion/recorder/HybridClarionRecorder.kt
package com.margelo.nitro.clarion.recorder

import android.content.Context
import com.clarionhq.recorder.RecorderCallbacks
import com.clarionhq.recorder.RecorderConfig
import com.clarionhq.recorder.RecorderError
import com.clarionhq.recorder.RecorderResult
import com.clarionhq.recorder.RecorderSession
import com.facebook.proguard.annotations.DoNotStrip
import com.margelo.nitro.NitroModules
import com.margelo.nitro.core.Promise
import java.util.concurrent.atomic.AtomicInteger

@DoNotStrip
class HybridClarionRecorder : HybridClarionRecorderSpec() {
  override val memorySize: Long get() = 0L

  private val context: Context = NitroModules.applicationContext
    ?: error("Nitro applicationContext is null — was the module initialized?")

  private val nextListenerId = AtomicInteger(1)
  private val stateListeners = mutableMapOf<Int, (String) -> Unit>()
  private val audioLevelListeners = mutableMapOf<Int, (Double, Double) -> Unit>()
  private val chunkListeners = mutableMapOf<Int, (String, Double, Double, Double) -> Unit>()
  private val errorListeners = mutableMapOf<Int, (NativeRecorderError) -> Unit>()

  @Volatile private var session: RecorderSession? = null
  @Volatile private var currentState: String = "idle"

  private val callbacks = object : RecorderCallbacks {
    override fun onState(state: String) {
      currentState = state
      stateListeners.values.toList().forEach { runCatching { it(state) } }
    }
    override fun onAudioLevel(rms: Double, peak: Double) {
      audioLevelListeners.values.toList().forEach { runCatching { it(rms, peak) } }
    }
    override fun onChunk(uri: String, startMs: Long, endMs: Long, sizeBytes: Long) {
      chunkListeners.values.toList().forEach {
        runCatching { it(uri, startMs.toDouble(), endMs.toDouble(), sizeBytes.toDouble()) }
      }
    }
    override fun onError(error: RecorderError) {
      currentState = "error"
      val payload = NativeRecorderError(
        code = error.code,
        message = error.message ?: "Unknown error",
        recoverable = error.recoverable,
      )
      errorListeners.values.toList().forEach { runCatching { it(payload) } }
    }
  }

  override val state: String get() = currentState

  override fun prepare(config: NativeRecorderConfig): Promise<Unit> =
    Promise.async {
      runOrThrow {
        val parsed = config.toKotlinConfig()
        val newSession = RecorderSession(context, parsed, callbacks)
        callbacks.onState("preparing")
        newSession.prepare()
        session = newSession
        callbacks.onState("ready")
      }
    }

  override fun start(): Promise<Unit> = Promise.async {
    runOrThrow {
      val s = session ?: throw RecorderError("ENGINE_NOT_READY", "Call prepare() first")
      callbacks.onState("starting")
      s.start()
    }
  }

  override fun pause(): Promise<Unit> = Promise.async {
    runOrThrow {
      session?.pause() ?: throw RecorderError("ENGINE_NOT_READY", "No session")
    }
  }

  override fun resume(): Promise<Unit> = Promise.async {
    runOrThrow {
      session?.resume() ?: throw RecorderError("ENGINE_NOT_READY", "No session")
    }
  }

  override fun stop(): Promise<NativeRecorderResult> = Promise.async {
    runOrThrow {
      val s = session ?: throw RecorderError("ENGINE_NOT_READY", "No session")
      val result = s.stop().toNative()
      session = null
      result
    }
  }

  override fun discard(): Promise<Unit> = Promise.async {
    runOrThrow {
      session?.discard()
      session = null
    }
  }

  override fun release(): Promise<Unit> = Promise.async {
    runOrThrow {
      session?.release()
      session = null
      removeAllListeners()
    }
  }

  private inline fun <T> runOrThrow(block: () -> T): T {
    return try {
      block()
    } catch (e: RecorderError) {
      throw RuntimeException("[${e.code}] ${e.message}", e)
    }
  }

  override fun addStateListener(listener: (String) -> Unit): Double =
    register(stateListeners, listener)

  override fun addAudioLevelListener(listener: (Double, Double) -> Unit): Double =
    register(audioLevelListeners, listener)

  override fun addChunkListener(
    listener: (String, Double, Double, Double) -> Unit,
  ): Double = register(chunkListeners, listener)

  override fun addErrorListener(listener: (NativeRecorderError) -> Unit): Double =
    register(errorListeners, listener)

  override fun removeListener(id: Double) {
    val key = id.toInt()
    stateListeners.remove(key)
    audioLevelListeners.remove(key)
    chunkListeners.remove(key)
    errorListeners.remove(key)
  }

  override fun removeAllListeners() {
    stateListeners.clear()
    audioLevelListeners.clear()
    chunkListeners.clear()
    errorListeners.clear()
  }

  private fun <T> register(map: MutableMap<Int, T>, listener: T): Double {
    val id = nextListenerId.getAndIncrement()
    map[id] = listener
    return id.toDouble()
  }
}

private fun NativeRecorderConfig.toKotlinConfig(): RecorderConfig = RecorderConfig(
  sampleRate = sampleRate.toInt(),
  channels = channels.toInt(),
  bitDepth = bitDepth.toInt(),
  outputDirectory = outputDirectory,
  filenamePrefix = filenamePrefix,
  rotateAfterMs = rotateAfterMs?.toLong(),
  emitAudioLevel = emitAudioLevel,
  audioLevelIntervalMs = audioLevelIntervalMs.toInt(),
  aacBitrate = aacBitrate.toInt(),
)

private fun RecorderResult.toNative(): NativeRecorderResult = NativeRecorderResult(
  uri = uri,
  durationMs = durationMs.toDouble(),
  sizeBytes = sizeBytes.toDouble(),
  sampleRate = sampleRate.toDouble(),
  channels = channels.toDouble(),
  bitDepth = bitDepth.toDouble(),
)
